modules/nix/nixpkgs.nix

```diff
@@ -44,72 +90,260 @@ let
+    system = mkOption {
       type = types.str;
       example = "x86_64-darwin";
+      default =
+        if opt.hostPlatform.isDefined
+        then
+          throw ''
+            Neither ${opt.system} nor any other option in nixpkgs.* is meant
+            to be read by modules and configurations.
+            Use pkgs.stdenv.hostPlatform instead.
+          ''
+        else
+          throw ''
+            Neither ${opt.hostPlatform} nor the legacy option ${opt.system} has been set.
+            The option ${opt.system} is still fully supported for interoperability,
+            but will be deprecated in the future, so we recommend to set ${opt.hostPlatform}.
+          '';
+      defaultText = lib.literalMD ''
+        Traditionally `builtins.currentSystem`, but unset when invoking nix-darwin through `lib.darwinSystem`.
+      '';
       description = lib.mdDoc ''
-        Specifies the Nix platform type for which NixOS should be built.
-        If unset, it defaults to the platform type of your host system.
-        Specifying this option is useful when doing distributed
-        multi-platform deployment, or when building virtual machines.
+        Specifies the Nix platform type on which nix-darwin should be built.
+        It is better to specify `nixpkgs.hostPlatform` instead.
+
+        Ignored when `nixpkgs.pkgs` or `nixpkgs.hostPlatform` is set.
+      '';
+    };
+
+    # nix-darwin only
+
+    source = mkOption {
+      type = types.path;
+      defaultText = literalMD ''
+        `<nixpkgs>` or nix-darwin's `nixpkgs` flake input
+      '';
+      description = lib.mdDoc ''
+        The path to import Nixpkgs from. If you're setting a custom
+        [](#opt-nixpkgs.pkgs) or `_module.args.pkgs`, setting this
+        to something with `rev` and `shortRev` attributes (such as a
+        flake input or `builtins.fetchGit` result) will also set
+        `system.nixpkgsRevision` and related options.
+        (nix-darwin only)
+      '';
+    };
+
+    constructedByUs = mkOption {
+      type = types.bool;
+      internal = true;
+      description = ''
+        Whether `pkgs` was constructed by this module. This is false when any of
+        `nixpkgs.pkgs` or `_module.args.pkgs` is set. (nix-darwin only)
       '';
     };
   };
 
   config = {
+    _module.args = {
+      pkgs =
+        # We explicitly set the default override priority, so that we do not need
+        # to evaluate finalPkgs in case an override is placed on `_module.args.pkgs`.
+        # After all, to determine a definition priority, we need to evaluate `._type`,
+        # which is somewhat costly for Nixpkgs. With an explicit priority, we only
+        # evaluate the wrapper to find out that the priority is lower, and then we
+        # don't need to evaluate `finalPkgs`.
+        lib.mkOverride defaultOverridePriority
+          finalPkgs.__splicedPackages;
+    };
 
-    # _module.args.pkgs is defined in ../../eval-config.nix
+    nixpkgs.constructedByUs =
```

Comment:
This is rather complicated to the point I can't really follow. Why does it matter where pkgs was set?

Reply:
This logic is directly from the upstream NixOS module; it is used to guard the assertions below. I exported it so that the version module can correctly determine whether it knows the origin of Nixpkgs and avoid setting incorrect version information as currently happens; if you override the package set then the `nixpkgs.source` defaults of nix-darwin's `nixpkgs` flake input or your `<nixpkgs>` channel have nothing to do with where `pkgs` came from or its version. Setting `nixpkgs.source` explicitly is available in those cases.

Reply:
cf.

https://github.com/LnL7/nix-darwin/blob/93f92130aec7d7f20d24bfa04b9a6c9be5fb4c9a/modules/system/version.nix#L20-L25

The consequences can be worse than just incorrect links in docs: I almost used it to gate the 22.11 docs backwards compatibility, which would have broken even more configurations than https://github.com/LnL7/nix-darwin/issues/718.

Reply:
Ah, I thought this was specific to the darwin module.

Reply:
WRT the version checking, yeah here you clearly notice (especially with flakes) that nixpkgs-lib and nixpkgs-pkgs should be separate packages.